src/svm/Elf.zig

```
@@ -0,0 +1,516 @@
+
+    self.dynamic_relocations_table = std.mem.bytesAsSlice(
+        elf.Elf64_Rel,
+        self.bytes[offset..][0..size],
+    );
+}
+
+fn parseDynamicSymbolTable(self: *Elf) !void {
+    const vaddr = self.dynamic_table[elf.DT_SYMTAB];
+    if (vaddr == 0) return;
+
+    for (self.shdrs, 0..) |shdr, i| {
+        if (shdr.sh_addr != vaddr) continue;
+
+        if (shdr.sh_type != elf.SHT_SYMTAB and shdr.sh_type != elf.SHT_DYNSYM) {
+            return error.InvalidSectionHeader;
+        }
+
+        self.dynamic_symbol_table = std.mem.bytesAsSlice(
+            elf.Elf64_Sym,
+            self.shdrSlice(@intCast(i)),
+        );
+        return;
+    } else return error.InvalidDynamicSectionTable;
+}
+
+pub fn parseRoSections(self: *const Elf, allocator: std.mem.Allocator) !Executable.Section {
+    const ro_names: []const []const u8 = &.{
+        ".text",
+        ".rodata",
+        ".data.rel.ro",
+        // ".eh_frame",
+    };
+
+    var lowest_addr: usize = std.math.maxInt(usize);
+    var highest_addr: usize = 0;
+
+    var ro_fill_length: usize = 0;
+    var invalid_offsets: bool = false;
+
+    var first_ro_section: usize = 0;
+    var last_ro_section: usize = 0;
+    var n_ro_sections: usize = 0;
+
+    var ro_slices = try std.ArrayListUnmanaged(struct {
+        usize,
+        []const u8,
+    }).initCapacity(allocator, self.shdrs.len);
+    defer ro_slices.deinit(allocator);
+
+    for (self.shdrs, 0..) |shdr, i| {
+        const name = self.getString(shdr.sh_name);
+        for (ro_names) |ro_name| {
+            if (std.mem.eql(u8, ro_name, name)) break;
+        } else continue;
+
+        if (n_ro_sections == 0) {
+            first_ro_section = i;
+        }
+        last_ro_section = i;
+        n_ro_sections = n_ro_sections +| 1;
+
+        const section_addr = shdr.sh_addr;
+
+        if (!invalid_offsets) {
+            if (section_addr != shdr.sh_offset) invalid_offsets = true;
```

Is it actually invalid if the address and ofset are not equal? Doesn't that just mean the ELF file itself is not addressed starting at 0? Is that a problem?

Also, invalid_offsets is never used for anything consequential. What's the point of this code?

---

Yes, it is invalid in V1. I don't understand what you mean by "addressed starting at 0" though.

And regarding `invalid_offsets`, that's a bit of logic that's used in the V2 branch, I forgot to remove it here.  Removing it now would just cause conflicts.

---

If it's invalid, then why not return an error here?

---

> If it's invalid, then why not return an error here?

My bad for making this a bit confusing, I'm just going to remove that boolean from this branch. It does return an error in the V2 branch.

> You're saying the address and offset are expected to be the same. Then why are they two separate fields?

They need to be the same in V1. In V2 they can be different values. It's easier for me to just implement this like I'm getting ready for V2, and not include the complicated V2 stuff.

> What is does addr mean? 
...
I was thinking of offset as meaning the number of bytes from the beginning of the file.

`addr` is the virtual address of the section in memory. You're correct that `offset` is the offset of the section into the file image.
We're checking them to be equal here, so it will be a value less than the size of the file. `highest_address` tracks the largest value we'll need to allocate to create this new virtual memory, and we return `ValueOutOfBounds` if that value is too large, i.e larger than the file image.

---

Removed the boolean here: 6394675672c593bd0845745f6bc65cc2f21a9ca7